File: src/services/MetricsService/MetricsProvider.ts
import {Metric} from './Metric';
import {MetricsStorageReader} from './MetricsStorage';

export interface MetricsProvider {
  retrieveAll(): Promise<Metric[]>;
  retrieveLaterThanTimestamp(timestamp: number): Promise<Metric[]>;
}

export class DefaultMetricsProvider implements MetricsProvider {
  private metricsStorage: MetricsStorageReader;

  constructor(metricsStorage: MetricsStorageReader) {
    this.metricsStorage = metricsStorage;
  }

  retrieveAll(): Promise<Metric[]> {
    return this.metricsStorage.retrieve();
  }

  retrieveLaterThanTimestamp(timestamp: number): Promise<Metric[]> {
    return this.retrieveAll().then(metrics => metrics.filter(metric => metric.timestamp > timestamp));
  }
}

File: android/app/src/main/java/app/covidshield/services/metrics/MetricsProvider.kt
package app.covidshield.services.metrics

interface MetricsProvider {
    suspend fun retrieveAll(): List<Metric>
    suspend fun retrieveLaterThanTimestamp(timestamp: Number): List<Metric>
}

class DefaultMetricsProvider(private val metricsStorage: MetricsStorageReader) : MetricsProvider {

    override suspend fun retrieveAll(): List<Metric> {
        return metricsStorage.retrieve()
    }

    override suspend fun retrieveLaterThanTimestamp(timestamp: Number): List<Metric> {
        return retrieveAll().filter { it.timestamp.toLong() > timestamp.toLong() }
    }

}